build: issues with configure on Power systems with PGI compilers

When I run configure on Power, with the MPICH 3.2.1 tarball available from the MPICH site. I get a failure at:
configure: error: unable to determine matching C type for C++ bool

I used the following configure line

./configure --prefix=$PWD/install-dir CC='pgcc' CXX='pgc++' F77='pgfortran' FC='pgfortran' 

Although I used PGI 17.10, the version of PGI compiler does not matter.

When I look at the config.log, the errors look like:

"/usr/include/powerpc64le-linux-gnu/sys/types.h", line 194: error: invalid
          combination of type specifiers
  __intN_t (8, __QI__);
  ^

"/usr/include/powerpc64le-linux-gnu/sys/types.h", line 194: error: an attribute
          specifies a mode incompatible with "<error-type>"
  __intN_t (8, __QI__);
  ^

The error is caused by the "#define int8_t signed char" line in conftest.cpp generated by the configure script.  The configure script seems to be looking for type int8_t using the macro “AC_TYPE_INT8_T”. It looks like it is not finding it and thus adding that line.

The conftest.cpp at the end of the ticket (sorry about that, the system did not allow me to attach a *.cpp file for some reason) built as follows will reproduce the issue

pgc++ -o conftest conftest.cpp

"/usr/include/sys/types.h", line 194: error: invalid combination of type
          specifiers
  __intN_t (8, __QI__);
  ^

"/usr/include/sys/types.h", line 194: error: an attribute specifies a mode
          incompatible with "<error-type>"
  __intN_t (8, __QI__);
  ^

"conftest.cpp", line 190: warning: statement is unreachable
    return 0;
    ^

```
/* confdefs.h */
#define PACKAGE_NAME "MPICH"
#define PACKAGE_TARNAME "mpich"
#define PACKAGE_VERSION "3.2.1"
#define PACKAGE_STRING "MPICH 3.2.1"
#define PACKAGE_BUGREPORT "discuss@mpich.org"
#define PACKAGE_URL "http://www.mpich.org/"
#define PACKAGE "mpich"
#define VERSION "3.2.1"
#define STDC_HEADERS 1
#define HAVE_SYS_TYPES_H 1
#define HAVE_SYS_STAT_H 1
#define HAVE_STDLIB_H 1
#define HAVE_STRING_H 1
#define HAVE_MEMORY_H 1
#define HAVE_STRINGS_H 1
#define HAVE_INTTYPES_H 1
#define HAVE_STDINT_H 1
#define HAVE_UNISTD_H 1
#define HAVE_DLFCN_H 1
#define LT_OBJDIR ".libs/"
#define HAVE_ERROR_CHECKING MPID_ERROR_LEVEL_ALL
#define MPICH_ERROR_MSG_LEVEL MPICH_ERROR_MSG_ALL
#define USE_LOGGING MPID_LOGGING_NONE
#define MPICH_THREAD_LEVEL MPI_THREAD_MULTIPLE
#define MPICH_IS_THREADED 1
#define MPICH_THREAD_GRANULARITY MPICH_THREAD_GRANULARITY_GLOBAL
#define MPIU_HANDLE_ALLOCATION_METHOD MPIU_HANDLE_ALLOCATION_MUTEX
#define MPIU_THREAD_REFCOUNT MPIU_REFCOUNT_NONE
#define HAVE_ROMIO 1
#define HAVE__FUNC__ /**/
#define HAVE__FUNCTION__ /**/
#define HAVE_PRAGMA_WEAK 1
#define USE_WEAK_SYMBOLS 1
#define HAVE_MULTIPLE_PRAGMA_WEAK 1
#define ENABLE_PVAR_NEM 0
#define ENABLE_PVAR_RECVQ 0
#define ENABLE_PVAR_RMA 0
#define HAVE_LONG_LONG 1
#define STDCALL
#define F77_NAME_LOWER_USCORE 1
#define HAVE_MPI_F_INIT_WORKS_WITH_C 1
#define HAVE_FORTRAN_BINDING 1
#define HAVE_F08_BINDING 0
#define TRUE 1
#define FALSE 0
#define HAVE_CXX_EXCEPTIONS /**/
#define HAVE_NAMESPACES /**/
#define HAVE_NAMESPACE_STD /**/
#define HAVE_CXX_BINDING 1
#define HAVE_NAMEPUB_SERVICE 1
#define restrict __restrict
#define HAVE_BUILTIN_EXPECT 1
#define WORDS_LITTLEENDIAN 1
#define HAVE_LONG_DOUBLE 1
#define HAVE_LONG_LONG_INT 1
#define HAVE_MAX_INTEGER_ALIGNMENT 8
#define HAVE_MAX_STRUCT_ALIGNMENT 8
#define HAVE_MAX_FP_ALIGNMENT 16
#define HAVE_MAX_DOUBLE_FP_ALIGNMENT 8
#define HAVE_MAX_LONG_DOUBLE_FP_ALIGNMENT 16
#define SIZEOF_CHAR 1
#define SIZEOF_UNSIGNED_CHAR 1
#define SIZEOF_SHORT 2
#define SIZEOF_UNSIGNED_SHORT 2
#define SIZEOF_INT 4
#define SIZEOF_UNSIGNED_INT 4
#define SIZEOF_LONG 8
#define SIZEOF_UNSIGNED_LONG 8
#define SIZEOF_LONG_LONG 8
#define SIZEOF_UNSIGNED_LONG_LONG 8
#define SIZEOF_FLOAT 4
#define SIZEOF_DOUBLE 8
#define SIZEOF_LONG_DOUBLE 16
#define SIZEOF_VOID_P 8
#define STDC_HEADERS 1
#define HAVE_STDDEF_H 1
#define SIZEOF_WCHAR_T 4
#define SIZEOF_FLOAT_INT 8
#define SIZEOF_DOUBLE_INT 16
#define SIZEOF_LONG_INT 16
#define SIZEOF_SHORT_INT 8
#define SIZEOF_TWO_INT 8
#define SIZEOF_LONG_DOUBLE_INT 32
#define HAVE_SYS_BITYPES_H 1
#define HAVE_INTTYPES_H 1
#define HAVE_STDINT_H 1
#define int8_t signed char
#define HAVE_INT8_T 1
#define HAVE_INT16_T 1
#define HAVE_INT32_T 1
#define HAVE_INT64_T 1
#define HAVE_UINT8_T 1
#define HAVE_UINT16_T 1
#define HAVE_UINT32_T 1
#define HAVE_UINT64_T 1
#define HAVE_STDBOOL_H 1
#define HAVE_COMPLEX_H 1
#define SIZEOF__BOOL 1
#define SIZEOF_FLOAT__COMPLEX 8
#define SIZEOF_DOUBLE__COMPLEX 16
#define SIZEOF_LONG_DOUBLE__COMPLEX 32
#define HAVE__BOOL 1
#define HAVE_FLOAT__COMPLEX 1
#define HAVE_DOUBLE__COMPLEX 1
#define HAVE_LONG_DOUBLE__COMPLEX 1
#define MPIR_REAL4_CTYPE float
#define MPIR_REAL8_CTYPE double
#define MPIR_REAL16_CTYPE long double
#define MPIR_INTEGER1_CTYPE char
#define MPIR_INTEGER2_CTYPE short
#define MPIR_INTEGER4_CTYPE int
#define MPIR_INTEGER8_CTYPE long
#define SIZEOF_F77_INTEGER 4
#define SIZEOF_F77_REAL 4
#define SIZEOF_F77_DOUBLE_PRECISION 8
#define MPIR_FC_REAL_CTYPE float
#define MPIR_FC_DOUBLE_CTYPE double
#define HAVE_AINT_LARGER_THAN_FINT 1
#define HAVE_AINT_DIFFERENT_THAN_FINT 1
#define HAVE_FINT_IS_INT 1
#define F77_TRUE_VALUE_SET 1
#define F77_TRUE_VALUE -1
#define F77_FALSE_VALUE 0
/* end confdefs.h.  */
#include <stdio.h>
#ifdef HAVE_SYS_TYPES_H
# include <sys/types.h>
#endif
#ifdef HAVE_SYS_STAT_H
# include <sys/stat.h>
#endif
#ifdef STDC_HEADERS
# include <stdlib.h>
# include <stddef.h>
#else
# ifdef HAVE_STDLIB_H
#  include <stdlib.h>
# endif
#endif
#ifdef HAVE_STRING_H
# if !defined STDC_HEADERS && defined HAVE_MEMORY_H
#  include <memory.h>
# endif
# include <string.h>
#endif
#ifdef HAVE_STRINGS_H
# include <strings.h>
#endif
#ifdef HAVE_INTTYPES_H
# include <inttypes.h>
#endif
#ifdef HAVE_STDINT_H
# include <stdint.h>
#endif
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
static long int longval () { return (long int) (sizeof (bool)); }
static unsigned long int ulongval () { return (long int) (sizeof (bool)); }
#include <stdio.h>
#include <stdlib.h>
int
main ()
{

  FILE *f = fopen ("conftest.val", "w");
  if (! f)
    return 1;
  if (((long int) (sizeof (bool))) < 0)
    {
      long int i = longval ();
      if (i != ((long int) (sizeof (bool))))
        return 1;
      fprintf (f, "%ld", i);
    }
  else
    {
      unsigned long int i = ulongval ();
      if (i != ((long int) (sizeof (bool))))
        return 1;
      fprintf (f, "%lu", i);
    }
  /* Do not output a trailing newline, as this causes \r\n confusion
     on some platforms.  */
  return ferror (f) || fclose (f) != 0;

  ;
  return 0;
}
```
Thx,
Hari.

--- a/configure.ac
+++ b/configure.ac
@@ -2802,10 +2802,6 @@ AC_CHECK_ALIGNOF(long long)
 AC_CHECK_ALIGNOF(long double)
 AC_CHECK_ALIGNOF(short)
 AC_CHECK_ALIGNOF(int)
-AC_CHECK_ALIGNOF(int8_t)
-AC_CHECK_ALIGNOF(int16_t)
-AC_CHECK_ALIGNOF(int32_t)
-AC_CHECK_ALIGNOF(int64_t)
 AC_CHECK_ALIGNOF(bool)
 AC_CHECK_ALIGNOF(wchar_t)
 
@@ -2816,50 +2812,44 @@ AC_DEFINE(HAVE_SYS_BITYPES_H,1,[Define if sys/bitypes.h exists])])
 # A C99 compliant compiler should have inttypes.h for fixed-size int types
 AC_CHECK_HEADERS(inttypes.h stdint.h)
 
-# Check for types
-AC_TYPE_INT8_T
-AC_TYPE_INT16_T
-AC_TYPE_INT32_T
-AC_TYPE_INT64_T
+AC_CHECK_SIZEOF(int8_t)
+AC_CHECK_SIZEOF(int16_t)
+AC_CHECK_SIZEOF(int32_t)
+AC_CHECK_SIZEOF(int64_t)
 
-# Temporary issue in autoconf integer type checking (remove when
-# autoconf fixes this or provides a workaround for it)
-if test "$ac_cv_c_int8_t" != no ; then
+AC_CHECK_SIZEOF(uint8_t)
+AC_CHECK_SIZEOF(uint16_t)
+AC_CHECK_SIZEOF(uint32_t)
+AC_CHECK_SIZEOF(uint64_t)
+
+AC_CHECK_ALIGNOF(int8_t)
+AC_CHECK_ALIGNOF(int16_t)
+AC_CHECK_ALIGNOF(int32_t)
+AC_CHECK_ALIGNOF(int64_t)
+
+if test "$ac_cv_sizeof_int8_t" -eq 1 ; then
     AC_DEFINE(HAVE_INT8_T,1,[Define if int8_t is supported by the C compiler])
 fi
-if test "$ac_cv_c_int16_t" != no ; then
+if test "$ac_cv_sizeof_int16_t" -eq 2 ; then
     AC_DEFINE(HAVE_INT16_T,1,[Define if int16_t is supported by the C compiler])
 fi
-if test "$ac_cv_c_int32_t" != no ; then
+if test "$ac_cv_sizeof_int32_t" -eq 4 ; then
     AC_DEFINE(HAVE_INT32_T,1,[Define if int32_t is supported by the C compiler])
 fi
-if test "$ac_cv_c_int64_t" != no ; then
+if test "$ac_cv_sizeof_int64_t" -eq 8 ; then
     AC_DEFINE(HAVE_INT64_T,1,[Define if int64_t is supported by the C compiler])
 fi
 
-# The following make these definitions:
-#   define _UINT<n>_T 1
-# if uint<n>_t is available.  E.g., define _UINT8_T as 1 if uint8_t is available
-# if not available, define uint<n>_t as the related C type, e.g.,
-#   define uint8_t unsigned char
-#
-AC_TYPE_UINT8_T
-AC_TYPE_UINT16_T
-AC_TYPE_UINT32_T
-AC_TYPE_UINT64_T
-
-# Temporary issue in autoconf integer type checking (remove when
-# autoconf fixes this or provides a workaround for it)
-if test "$ac_cv_c_uint8_t" != no ; then
+if test "$ac_cv_sizeof_uint8_t" -eq 1 ; then
     AC_DEFINE(HAVE_UINT8_T,1,[Define if uint8_t is supported by the C compiler])
 fi
-if test "$ac_cv_c_uint16_t" != no ; then
+if test "$ac_cv_sizeof_uint16_t" -eq 2 ; then
     AC_DEFINE(HAVE_UINT16_T,1,[Define if uint16_t is supported by the C compiler])
 fi
-if test "$ac_cv_c_uint32_t" != no ; then
+if test "$ac_cv_sizeof_uint32_t" -eq 4 ; then
     AC_DEFINE(HAVE_UINT32_T,1,[Define if uint32_t is supported by the C compiler])
 fi
-if test "$ac_cv_c_uint64_t" != no ; then
+if test "$ac_cv_sizeof_uint64_t" -eq 8 ; then
     AC_DEFINE(HAVE_UINT64_T,1,[Define if uint64_t is supported by the C compiler])
 fi
 

